chore:(docs) Adding Jest testing document (#31554)
microsoft/fluentui: docs/react-v9/contributing/testing-with-jest.md

> > 🚨🚨 **This page is primarily about `@fluentui/react-components` ("v9") and `@fluentui/react` ("v8") and related packages.** 🚨🚨&nbsp; [See this page for `@fluentui/react-northstar` ("v0").](https://github.com/microsoft/fluentui/blob/master/packages/fluentui/test-a-feature.md)

Reviewer: ## [🕵 fluentui-web-components-v3 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=31554&commitId=c37b8a0c1affb48b4f3de25844a82c7530349d38&env=prod&clientType=FLUENTUI)

  <details><summary><b>Accordion</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Accordion.Size.normal.chromium.png|2| Changed|
|Accordion.Size - Dark Mode.normal.chromium.png|2| Changed|
|Accordion.Size - RTL.normal.chromium.png|2| Changed|

</details>

  <details><summary><b>Avatar</b> 5 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar.Color - RTL.normal.chromium.png|1| Changed|
|Avatar.Active Appearance.normal.chromium.png|10| Changed|
|Avatar.Active.normal.chromium.png|5| Changed|
|Avatar.Color.normal.chromium.png|1| Changed|
|Avatar.Active Appearance - Dark Mode.normal.chromium.png|2| Changed|

</details>

  <details><summary><b>Badge</b> 6 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Badge.Appearance.normal.chromium.png|1| Changed|
|Badge.Shape.normal.chromium.png|1| Changed|
|Badge.Color.normal.chromium.png|1| Changed|
|Badge.Appearance - Dark Mode.normal.chromium.png|15| Changed|
|Badge.Size.normal.chromium.png|6| Changed|
|Badge.Default.normal.chromium.png|1| Changed|

</details>

  <details><summary><b>Button</b> 15 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Button.Large Icon Only - Dark Mode.hover.chromium.png|3| Changed|
|Button.Large Icon Only.hover.chromium.png|4| Changed|
|Button.Large Icon Only.default.chromium.png|5| Changed|
|Button.Large With Icon - RTL.hover.chromium.png|4| Changed|
|Button.Large Icon Only - Dark Mode.default.chromium.png|4| Changed|
|Button.Large Icon Only - Dark Mode.pressed.chromium.png|5| Changed|
|Button.Large With Icon - Dark Mode.pressed.chromium.png|5| Changed|
|Button.Large With Icon - RTL.pressed.chromium.png|3| Changed|
|Button.Large With Icon - RTL.default.chromium.png|5| Changed|
|Button.Large With Icon - Dark Mode.hover.chromium.png|3| Changed|
|Button.Large With Icon.hover.chromium.png|4| Changed|
|Button.Large Icon Only.pressed.chromium.png|3| Changed|
|Button.Large With Icon.pressed.chromium.png|3| Changed|
|Button.Large With Icon - Dark Mode.default.chromium.png|4| Changed|
|Button.Large With Icon.default.chromium.png|5| Changed|

</details>

  <details><summary><b>Divider</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Divider.With Svg.normal.chromium.png|3| Changed|
|Divider.With Svg - RTL.normal.chromium.png|3| Changed|
|Divider.Vertical With Svg.normal.chromium.png|4| Changed|

</details>

  <details><summary><b>MenuList</b> 16 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|MenuList.Checkbox With Icons.1st selected.chromium.png|4| Changed|
|MenuList.Checkbox With Icons.normal.chromium.png|4| Changed|
|MenuList.Checkbox With Icons - RTL.2nd selected.chromium.png|5| Changed|
|MenuList.Radio With Icons - RTL.normal.chromium.png|4| Changed|
|MenuList.Checkbox With Icons - RTL.1st selected.chromium.png|4| Changed|
|MenuList.Checkbox - RTL.2nd selected.chromium.png|1| Changed|
|MenuList.Checkbox With Icons.2nd selected.chromium.png|5| Changed|
|MenuList.Radio With Icons.normal.chromium.png|4| Changed|
|MenuList.Checkbox With Icons - RTL.normal.chromium.png|4| Changed|
|MenuList.Radio With Icons.1st selected.chromium.png|4| Changed|
|MenuList.Checkbox.2nd selected.chromium.png|1| Changed|
|MenuList.Radio With Icons - RTL.1st selected.chromium.png|4| Changed|
|MenuList.Radio With Icons.2nd selected.chromium.png|4| Changed|
|MenuList.Radio With Icons - RTL.2nd selected.chromium.png|4| Changed|
|MenuList.With Icons.hover menuitem.chromium.png|3| Changed|
|MenuList.With Icons.normal.chromium.png|4| Changed|

</details>

  <details><summary><b>Slider</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Slider.Size Small.normal.chromium.png|1| Changed|
|Slider.Size Small.rightArrow.chromium.png|1| Changed|

</details>

  <details><summary><b>Switch</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Switch.Checked.hover.chromium.png|1| Changed|
|Switch.Checked - RTL.hover.chromium.png|1| Changed|

</details>

  <details><summary><b>Text</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Text.Block.normal.chromium.png|995| Changed|

</details>

  <details><summary><b>TextInput</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|TextInput.Size Large.normal.chromium.png|2| Changed|

</details>

<div id="vrtCommentfluent-ui_VRT_Pipeline_web-components"/>

> > 🚨🚨 **This page is primarily about `@fluentui/react-components` ("v9") and `@fluentui/react` ("v8") and related packages.** 🚨🚨&nbsp; [See this page for `@fluentui/react-northstar` ("v0").](https://github.com/microsoft/fluentui/blob/master/packages/fluentui/test-a-feature.md)

Reviewer: ## [🕵 fluentuiv8 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=31554&commitId=c37b8a0c1affb48b4f3de25844a82c7530349d38&env=prod&clientType=FLUENTUI)

  <details><summary><b>Pivot - Overflow</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Pivot - Overflow.Root.Narrow - Last tab selected.chromium.png|0| Removed|
|Pivot - Overflow.Root.Narrow - Overflow menu.chromium.png|0| Removed|

</details>

<div id="vrtCommentfluent-ui_VRT_Pipeline_v8"/>

> > 🚨🚨 **This page is primarily about `@fluentui/react-components` ("v9") and `@fluentui/react` ("v8") and related packages.** 🚨🚨&nbsp; [See this page for `@fluentui/react-northstar` ("v0").](https://github.com/microsoft/fluentui/blob/master/packages/fluentui/test-a-feature.md)

Reviewer: ## [🕵 fluentuiv9 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=31554&commitId=c37b8a0c1affb48b4f3de25844a82c7530349d38&env=prod&clientType=FLUENTUI)

  <details><summary><b>Card Converged - Interactive</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Card Converged - Interactive.appearance interactive - Filled Alternative - Dark Mode.click.chromium.png|0| Removed|
|Card Converged - Interactive.appearance interactive - Filled Alternative - Dark Mode.hover.chromium.png|0| Removed|
|Card Converged - Interactive.appearance interactive - Filled Alternative - Dark Mode.normal.chromium.png|0| Removed|

</details>

  <details><summary><b>Card Converged - Selectable</b> 4 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Card Converged - Selectable.appearance selectable - Subtle.click.chromium.png|0| Removed|
|Card Converged - Selectable.appearance selectable - Subtle.hover.chromium.png|0| Removed|
|Card Converged - Selectable.appearance selectable - Subtle.normal.chromium.png|0| Removed|
|Card Converged - Selectable.appearance selectable - Subtle.selected.chromium.png|0| Removed|

</details>

<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

## Overview

> Fluent UI's unit, functional, and snapshot tests are built using [Jest](https://jestjs.io/). This allows us to run tests in a Node environment but simulates the browser using [jsdom](https://github.com/jsdom/jsdom). ([See this page for other types of tests.](Testing))

Reviewer 2: dead link `([See this page for other types of tests.](Testing))`

We use various libraries on top of Jest for rendering and interacting with React components:

> - New tests should use [React Testing Library](https://testing-library.com/docs/react-testing-library/intro)

Reviewer 2: should doesn't imply the testing approach , just list RTL 

```suggestion
- [React Testing Library](https://testing-library.com/docs/react-testing-library/intro)
```

> - Legacy tests may also use:
>   - [Enzyme](https://enzymejs.github.io/enzyme/) -- **not allowed in new tests**
>   - [react-test-renderer](https://reactjs.org/docs/test-renderer.html), primarily for snapshot testing _(snapshots are discouraged in new tests)_

Reviewer 2: Let's keep this as "FYI" to not confuse contributors that this is something we use as a standard

```suggestion
> [!NOTE]  
> Old monorepo libraries (like v8)( **non of these approaches is allowed in any new tests** ):
> - [Enzyme](https://enzymejs.github.io/enzyme/) -- **not allowed in new tests**
> - [react-test-renderer](https://reactjs.org/docs/test-renderer.html), primarily for snapshot testing _(snapshots are discouraged in new tests)_
```

## Running tests

> Our Jest setup generally require that packages be built before testing, so before running tests the first time, you should run `yarn build --to my-package` from the repo root.

Reviewer 2: ```suggestion

From your Terminal invoke one of following commands:

`yarn nx run <package-name>:test` or `yarn lage test --to <package-name>`

> [!NOTE] 
> New packages (v9) don't require build step before, which is not true for old fluent libraries like v8. 
Until v8 lives within our main branch we cannot remove `build` dependency from target task graph thus you'll experience build step even for new packages.
> 
> You can use `yarn workspace <package-name> test` for new packages if you don't have build within your cache and don't wanna wait for `build` step
```

> To run the tests for one package, `cd` to that package and then run `yarn test`.

Reviewer 2: This wont work in the future, and it's an antipattern, lets remove this
```suggestion
```

> To run an individual test (or technically any tests with a path matching this substring), `cd` to the relevant package and run `yarn jest MyTestName`.

Reviewer 2: pls remove, and point folks to jest documentation. In general I'd suggest that contributors use `watch` mode where they can narrow down test run from CLI - betrer DX

```suggestion
```

> (You can also run tests for the whole repo by running `yarn test` at the repo root. This will build beforehand if needed.)

Reviewer 2: redundant info which besides v-build team probably nobody uses/should do

```suggestion
```

## Running tests in watch mode

When you are developing tests, use the watch mode to run the tests as you write them!

> 1. Go to the package folder where you want to run the tests.

Reviewer 2: Lets' stop encouraging users to use `cd` in order to run tasks - this will not work in future

```suggestion
```

2. The start command varies by library: for v8, `yarn start-test`, or for v9, `yarn test --watch`
3. Edit and saving tests should now cause the console to re-run the tests you have added/modified.

### Debugging

The repo includes launch configurations for debugging tests using Visual Studio Code. (You could also configure debugging in another editor of your choice.)

> 1. Set breakpoints in the test file (`*.test.ts` or `*.test.tsx`)

Reviewer 2: ```suggestion
1. Set breakpoints in the test file (`*.test.ts`,`*.spec.ts`,`*.test.tsx`,`*.spec.tsx`)
```

2. Open the **Run** (debugger) pane in the sidebar and choose the configuration you want: usually **Debug current open test** to run only the current open test, or **Debug test** to run all tests for the package the current file is in
3. Start debugging

## Jest basics

> Tests in Jest are written similarly to Mocha tests, though Jest includes a number of assertions that work similarly to Chai. (If you've never used either of those, don't worry about it!)

Reviewer 2: I'm not sure any of these information is necessary as this document doesn't provide info about what is mocha nor chai. 

Lets simplify, make it concise and to the point.

```suggestion
```

- [Jest API reference](https://jestjs.io/docs/api)
  - [Available `expect` matchers](https://jestjs.io/docs/expect)
  - [Mock functions](https://jestjs.io/docs/mock-function-api)
  - [The `jest` object](https://jestjs.io/docs/jest-object) including fake timer APIs ([more info](https://jestjs.io/docs/timer-mocks))

Note that you do not need to import the assertions or the Jest APIs; they should be available automatically through the included typings.

A basic test example:

```ts
describe('thing', () => {
  it('does something', () => {
    expect(thing()).toEqual(aValue);
  });
});
```

## Testing components

In cases where you need to automate a component and validate it performs correctly, you can use [React Testing Library](https://testing-library.com/docs/react-testing-library/intro) to mount components, evaluate DOM structure, and simulate events.

Here's a very basic example of what a test might look like (we'll go over this in more detail later):

```tsx
const onClick = jest.fn();
const { getByRole } = render(<Button onClick={onClick}>This is a button</Button>);

userEvent.click(getByRole('button'));
expect(onClick).toHaveBeenCalled();
```

> ### Philosophy

Reviewer 2: this whole chapter should be renamed to something related to v8

> The primary purpose of React Testing Library is to increase confidence in your tests by **testing your components in the way a user would use them.** Users don't care what happens behind the scenes, they just see and interact with the output. Instead of accessing the components' internal APIs or evaluating their state, you'll get more confidence by writing your tests based on the component output.

Some examples of this philosophy:

- Prefer [finding elements like a user would](https://testing-library.com/docs/queries/about/#priority) (by role, label, or text) rather than relying on `className`, `id`, or DOM structure. This approach helps encourage writing UI that's accessible to all users.
- Prefer interact with elements the way the user would, by clicking or using the keyboard, rather than using imperative APIs.
  - [`@testing-library/user-event`](https://testing-library.com/docs/ecosystem-user-event/) more realistically simulates the full sequence of events that occur during user interaction.

Testing Library's APIs also make it _easier_ to write realistic, user-focused tests compared to Enzyme.

### Recommended reading

> This page will only go into basics of using Testing Library (and some [common pitfalls](#tips-and-faq)).

Reviewer 2: ` (and some [common pitfalls](#tips-and-faq))` dead link

For a more detailed run-through of how to write tests, check out this [tutorial](https://www.robinwieruch.de/react-testing-library/) recommended in Testing Library's docs. (One difference: we typically use the queries returned by `render()`, rather than the global queries from `screen` like the tutorial uses.)

It's highly recommended that you check out Testing Library's docs as well!

> - Basics

Reviewer 2: I'm not sure if it;s good idea to copy paste outlines of existing docs, can we simplify this to one link ?

Author: I'm not sure what you mean here? Generally speaking, I like having a basic overview in a single doc that can help an intern understand something, and then link out to the big thick heavy manual.

Reviewer 2: All of this is already outlined and well documented within RTL docs 

<img width="391" alt="image" src="https://github.com/microsoft/fluentui/assets/1223799/98969208-1d52-4d96-9c27-4efca2f0f350">

having manually maintained copy of well documented docs for 3rd party tool is not worth the effort of keeping it up to date, one link (https://testing-library.com/docs/react-testing-library/intro) with some encouragement to explore all the chapters is everything one should need.

Author: Okay - I've removed the bulk of that and referred to the Jest and RTL docs.
